jl4-core/src/L4/Evaluate.hs

```diff
@@ -572,11 +588,21 @@ exception exc _stack = do
   throwError exc
 
 runBinOp :: BinOp -> Value -> Value -> Stack -> Eval Value
-runBinOp BinOpPlus   (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 + num2)
-runBinOp BinOpMinus  (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 - num2)
-runBinOp BinOpTimes  (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 * num2)
-runBinOp BinOpDividedBy (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 `div` num2)
-runBinOp BinOpModulo    (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 `mod` num2)
+runBinOp BinOpPlus      (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 + num2)
+runBinOp BinOpMinus     (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 - num2)
+runBinOp BinOpTimes     (ValNumber num1) (ValNumber num2) _stack = pure $ ValNumber (num1 * num2)
+runBinOp BinOpDividedBy (ValNumber num1) (ValNumber num2)  stack
+  | num2 /= 0 = pure $ ValNumber (num1 / num2)
+  | otherwise = exception DivisionByZero stack
+runBinOp BinOpModulo    (ValNumber num1) (ValNumber num2) stack  =
+  case isInteger num1 of
+    Nothing -> exception (NotAnInteger num1) stack
+    Just n1 -> case isInteger num2 of
+      Nothing -> exception (NotAnInteger num2) stack
+      Just n2 ->
+        if n2 /= 0
+          then pure $ ValNumber (toRational $ n1 `mod` n2)
+          else exception DivisionByZero stack
```

Comment:
This could likely be cleaned up with dedicated Eval helpers such as `expectNonZeroInteger` and `expectInteger`. Does that sounds like an improvement?